gossip/gossip.go

```diff
@@ -159,7 +159,20 @@ func New(rpcContext *rpc.Context, resolvers []resolver.Resolver) *Gossip {
 	}
 
 	// Add ourselves as a SystemConfig watcher.
-	g.is.registerCallback(KeySystemConfig, g.updateSystemConfig)
+	updateC, unregister := g.RegisterUpdateChannel(KeySystemConfig)
+	g.stopper.RunWorker(func() {
+		for {
+			select {
+			case n := <-updateC:
+				g.updateSystemConfig(n)
+				updateC <- n
```

Comment:
As long as it doesn't deadlock with an unbuffered channel, I'm ok with allowing the callbacks to block gossip (any callbacks that do anything expensive must do so asynchronously, but they can make their own decision about how exactly to do that). If the naive approach doesn't work then I'd use a dedicated callback-runner goroutine.

It's true that reading and writing the same channel is not necessarily unidiomatic. The thing that stood out to me in this case is that it works only because the channel is unbuffered - if a buffer were added to the channel you could go into a tight loop of reading and writing the same value.